pcicrawler/cli.py

```diff
@@ -65,6 +74,283 @@ def jsonify(dev, hexify=False, vpd=False, aer=False):
     return jd
 
 
+class ocp_output_obj(object):
+    def __init__(self,
+                 devs: Set[PCIDevice],
+                 json_file: str,
+                 ocp_run: tv.TestRun):
+        self._json_file = json_file
+        self._devs = devs
+        self._ocp_run = ocp_run
+
+    class OCPTestStep(object):
+        def __init__(self, device, expected_result, ocp_run):
+            self._device = device
+            self._expected_result = expected_result
+            self._ocp_run = ocp_run
+            self._error_messages = []
+
+        @abstractmethod
```

Comment:
I enforcing it because of each validation check requires to pass in it's own value it's comparing against in the PCIe device list

e.g.
    def run(self):
        return super().run(self._device.express_link.cur_speed) <- this is the field to validate the device against

Reply:
i dont follow your logic here; what is the relation between abstractmethod and the validations passing in own expected values?

Reply:
question remains, now in `ocp_diag.py:34`. an abstractmethod cannot have a body, it's no longer abstract. besides that your class is not abstract
please see docs https://docs.python.org/3/library/abc.html#abc.abstractmethod

Reply:
resolved; now using metaclass ABC